File: src/ruby/tormenta20/models/acessorio_magico.rb
# frozen_string_literal: true

require "active_record"

module Tormenta20
  module Models
    # Magic accessory items (rings, amulets, cloaks, etc.) with passive and active effects.
    #
    # @example
    #   AcessorioMagico.menores.all
    #   AcessorioMagico.find_by(id: "anel_de_protecao").efeito
    #
    # @!attribute [rw] id
    #   @return [String] Unique identifier
    # @!attribute [rw] name
    #   @return [String] Item name
    # @!attribute [rw] categoria
    #   @return [String] Power tier: +"menor"+, +"medio"+, or +"maior"+
    # @!attribute [rw] preco
    #   @return [Integer, nil] Price in tibares
    # @!attribute [rw] roll_min
    #   @return [Integer, nil] Minimum roll on random loot table
    # @!attribute [rw] roll_max
    #   @return [Integer, nil] Maximum roll on random loot table
    # @!attribute [rw] description
    #   @return [String, nil] Item description
    class AcessorioMagico < Base
      self.table_name = "acessorios_magicos"

      # @!group Scopes
      # @!method menores
      #   @return [ActiveRecord::Relation] Minor-tier accessories
      scope :menores, -> { where(categoria: "menor") }
      # @!method medios
      #   @return [ActiveRecord::Relation] Medium-tier accessories
      scope :medios,  -> { where(categoria: "medio") }
      # @!method maiores
      #   @return [ActiveRecord::Relation] Major-tier accessories
      scope :maiores, -> { where(categoria: "maior") }
      # @!endgroup

      # Returns the parsed effect hash, handling both raw JSON strings and pre-parsed hashes.
      # @return [Hash] Effect data, or +{}+ on parse failure
      def efeito
        raw = read_attribute(:efeito)
        return {} if raw.blank?

        raw.is_a?(Hash) ? raw : JSON.parse(raw)
      rescue JSON::ParserError
        {}
      end
    end
  end
end

File: src/ts/models/acessorio_magico.ts
import { BaseModel } from "./base.js"
import { Query } from "../query.js"

const TABLE = "acessorios_magicos"

/**
 * An Acessório Mágico (magic accessory) in Tormenta20.
 *
 * Includes rings, amulets, cloaks, and other slotted magic items.
 *
 * @example
 * ```ts
 * AcessorioMagico.all()
 * AcessorioMagico.menores().all()
 * AcessorioMagico.maiores().all()
 * AcessorioMagico.find("anel_de_protecao")
 * ```
 */
export class AcessorioMagico extends BaseModel {
  /** Type of accessory (e.g. `"anel"`, `"amuleto"`, `"capa"`). */
  get categoria(): string { return this._row.categoria as string }
  /** Price in tibares. */
  get preco(): number { return (this._row.preco as number) ?? 0 }
  /** Minimum roll on the random loot table, or `null`. */
  get rollMin(): number | null { return (this._row.roll_min as number) ?? null }
  /** Maximum roll on the random loot table, or `null`. */
  get rollMax(): number | null { return (this._row.roll_max as number) ?? null }
  get description(): string | null { return (this._row.description as string) ?? null }
  /** JSON object with mechanical effects. */
  get efeito(): Record<string, unknown> { return BaseModel.json(this._row.efeito, {}) }

  toH() {
    return {
      id: this.id, name: this.name, categoria: this.categoria, preco: this.preco,
      roll_min: this.rollMin, roll_max: this.rollMax,
      description: this.description, efeito: this.efeito,
    }
  }

  static query(): Query<AcessorioMagico> { return BaseModel.makeQuery(TABLE, AcessorioMagico) }
  static all(): AcessorioMagico[] { return AcessorioMagico.query().all() }
  static find(id: string): AcessorioMagico | null { return AcessorioMagico.query().find(id) }
  static first(): AcessorioMagico | null { return AcessorioMagico.query().first() }
  static last(): AcessorioMagico | null { return AcessorioMagico.query().last() }
  static count(): number { return AcessorioMagico.query().count() }

  static menores(): Query<AcessorioMagico> { return AcessorioMagico.query().where("categoria = ?", "menor") }
  static medios(): Query<AcessorioMagico> { return AcessorioMagico.query().where("categoria = ?", "medio") }
  static maiores(): Query<AcessorioMagico> { return AcessorioMagico.query().where("categoria = ?", "maior") }
}
